### ISSUE TITLE
Package from client to server gets lost sometimes - followed by reconnect

### ISSUE THREAD

**Issue by maikokuppe** [rails/actioncable#100]
I'm developing a websocket-based board game. Every time a user makes a move, it is sent to the server via websocket connection.
Sometimes a move is sent (the .perform method is called), some seconds pass, and there comes an automatic reconnect. The request was not sent. Instead, this appears in the log file where the incoming request should appear:

```
D, [2015-10-20T10:54:47.471470 #26807] DEBUG -- : Terminating 101 actors...
W, [2015-10-20T10:54:47.490726 #26807]  WARN -- : Terminating task: type=:finalizer, meta={:method_name=>:__shutdown__}, status=:receiving
	Celluloid::TaskFiber backtrace unavailable. Please try `Celluloid.task_class = Celluloid::TaskThread` if you need backtraces here.
```

I suppose, this is just an accidental disconnect, no one can prevent.
My suggestion is that all packages being sent through the websocket should auto-retry after some time if they did not receive a confirmation like 'I received package 927374!'

**Comment by kayakyakr** [comment]
That looks more like a server crash of some sort. It's shutting down Celluloid. Is there any additional information in the log?

Are you by chance using custom broadcast blocks? I was having issues with silent failures in custom blocks earlier.

**Comment by maikokuppe** [comment]
That log excerpt included everything that could possibly be related to that incident.

I don't know what you mean by 'custom broadcast blocks'. I broadcast user A's move to user B like this
```
ActionCable.server.broadcast "user_#{current_user.id}", { :action => 'move', :data => data }
```
But that doesn't happen here, since the server does not receive the move.

**Comment by kayakyakr** [comment]
The log is simply telling you that Celluloid is shutting down (`Terminating 101 actors`). So something is happening that is causing the server to go, "oh hey, this is bad" and exit, albeit gracefully with Celluloid taking the time to shut down its actors and trying to display backtraces.

I asked about custom block because you are able to do something like:

```ruby
stream_from 'channel', ->(msg) do
  # take some actions on message
  transmit msg
end
```

And I've found that it fails silently if something goes wrong inside that block.

Basically, what I'm saying is that it looks more like a server crash than a connection issue requiring resend and that you're going to have to hunt down why it's intermittently crashing. Are you running standalone or in-app? I'm going to assume that you've watched your websocket frames to ensure that the message has gone out from the client. You may need to set up debug breakpoints inside the channel and possibly even inside the gem in order to step through and see where things go wrong.

**Comment by maikokuppe** [comment]
I'm not using custom blocks and I'm running standalone. I do clearly remember the message being sent out from the client.

Unfortunately I cannot debug this well, since it happens very rarely. Once in 400 requests I'm estimating by the logs.

I agree, that it is probably not a connection issue: When I try to reproduce it by cutting the internet connection for some seconds, it automatically sends the request 5 seconds later, when the connection is reestablished.

**Comment by dhh** [comment]
Please reopen if you find a reproducible case. Thanks!
